Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,629 @@
+
+    @staticmethod
+    def _staggered_function(num_indices, stride, iter_number):
+        """Function that takes in an iteration number and outputs an index number based on the staggered ordering.  """
+        iter_number_mod = iter_number % num_indices
+        floor = num_indices // stride
+        mod = num_indices % stride
+
+        if iter_number_mod < (floor + 1)*mod:
+            row_number = iter_number_mod // (floor + 1)
+            column_number = (iter_number_mod % (floor + 1))
+        else:
+            row_number = mod + (iter_number_mod - (floor+1)*mod) // floor
+            column_number = (iter_number_mod - (floor+1)*mod) % floor
+
+        return row_number + stride*column_number
+
+    @staticmethod
+    def staggered(num_indices, stride):
+        """
+        Instantiates a sampler which outputs in a staggered order. 
+
+        Parameters
+        ----------
+        num_indices: int
+            One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+        stride: int
+            The sampler will output in the order {0, a, 2a, 3a, ...., 1, 1+a, 1+2a, 1+3a,...., 2, 2+a, 2+2a, 2+3a,...} where a=stride. 
+            The stride should be less than the num_indices
+
+        Returns
+        -------
+        A Sampler  that can be called with Sampler.next()  or next(Sampler) and outputs in a staggered ordering
+
+        Example
+        -------
+        >>> sampler=Sampler.staggered(21,4)
+        >>> print(sampler.get_samples(5))
+        [ 0  4  8 12 16]
+        """
+
+        if stride >= num_indices:
+            raise (ValueError('The stride should be less than the number of indices'))
+
+        sampler = Sampler(function=partial(Sampler._staggered_function, num_indices, stride), num_indices=num_indices, sampling_type='staggered', prob_weights=[
+            1/num_indices]*num_indices)
+
+        return sampler
+
+    @staticmethod
+    def random_with_replacement(num_indices, prob=None, seed=None):
+        """
+        Instantiates a sampler which outputs from a list of indices {0, 1, …, S-1}, with S = num_indices, with given probability and with replacement. 
+
+        Parameters
+        ----------
+        num_indices: int
+            One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S = num_indices. 
+
+        prob: list of floats of length num_indices that sum to 1. default is None
+            This is the probability for each index to be called by next. If None, then the indices will be sampled uniformly. 
+
+        seed:int, default is None
+            Random seed for the random number generator.  If set to None, the seed will be set using the current time.
+
+        Returns
+        -------
+        A Sampler  that can be called with Sampler.next()  or next(Sampler) that samples randomly with replacement 
```

Comment:
Returns
        -------
        RandomSampler
            An instance of the RandomSampler class that will generate indices XXXX